parachain/pallets/collab-ai/common/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+	Neutral,
+	/// Want this guardian no matter which pool proposal
+	#[codec(index = 1)]
+	Aye,
+	/// Against this guardian no matter which pool proposal
+	#[codec(index = 2)]
+	Nay,
+	/// Support this guardian for only specific pool proposal
+	/// And neutral for other pool proposal
+	#[codec(index = 3)]
+	Specific(PoolProposalIndex),
+}
+
+/// Some sort of check on the account is from some group.
+pub trait CuratorQuery<AccountId> {
+	/// All curator but banned ones
+	fn is_curator(account: AccountId) -> bool;
+
+	/// Only verified one
+	fn is_verified_curator(account: AccountId) -> bool;
+}
+
+pub trait ProposerQuery<AccountId> {
+	fn is_proposer(account: AccountId, proposal_index: PoolProposalIndex) -> bool;
+}
+
+pub struct EnsureSignedAndCurator<AccountId, EC>(PhantomData<(AccountId, EC)>);
+impl<
+		O: Into<Result<RawOrigin<AccountId>, O>> + From<RawOrigin<AccountId>>,
+		AccountId: Decode + Clone,
+		EC,
+	> EnsureOrigin<O> for EnsureSignedAndCurator<AccountId, EC>
+where
+	EC: CuratorQuery<AccountId>,
+{
+	type Success = AccountId;
+	fn try_origin(o: O) -> Result<Self::Success, O> {
+		o.into().and_then(|o| match o {
+			RawOrigin::Signed(who) => {
+				if EC::is_curator(who.clone()) {
+					Ok(who)
+				} else {
+					Err(O::from(RawOrigin::Signed(who)))
+				}
+			},
+			r => Err(O::from(r)),
+		})
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn try_successful_origin() -> Result<O, ()> {
+		// No promised successful_origin
+		Err(())
+	}
+}
+
+pub struct EnsureSignedAndVerifiedCurator<AccountId, EC>(PhantomData<(AccountId, EC)>);
+impl<
+		O: Into<Result<RawOrigin<AccountId>, O>> + From<RawOrigin<AccountId>>,
+		AccountId: Decode + Clone,
+		EC,
+	> EnsureOrigin<O> for EnsureSignedAndVerifiedCurator<AccountId, EC>
+where
+	EC: CuratorQuery<AccountId>,
+{
+	type Success = AccountId;
+	fn try_origin(o: O) -> Result<Self::Success, O> {
+		o.into().and_then(|o| match o {
+			RawOrigin::Signed(who) => {
+				if EC::is_verified_curator(who.clone()) {
+					Ok(who)
+				} else {
+					Err(O::from(RawOrigin::Signed(who)))
+				}
+			},
+			r => Err(O::from(r)),
+		})
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn try_successful_origin() -> Result<O, ()> {
+		// No promised successful_origin
+		Err(())
+	}
+}
+
+pub type EnsureRootOrVerifiedCurator<AccountId, EC> =
+	EitherOfDiverse<EnsureRoot<AccountId>, EnsureSignedAndVerifiedCurator<AccountId, EC>>;
+
+pub const INVESTING_POOL_INDEX_SHIFTER: u128 = 1_000_000_000_000_000;
+pub const INVESTING_POOL_START_EPOCH_SHIFTER: u128 = 1_000;
+pub const INVESTING_POOL_END_EPOCH_SHIFTER: u128 = 1;
+
+pub struct InvestingPoolAssetIdGenerator<AssetId>(PhantomData<AssetId>);
+impl<AssetId: From<u128> + Into<u128>> InvestingPoolAssetIdGenerator<AssetId> {
+	/// Create a series of new asset id based on pool index and reward epoch
+	/// Return None if impossible to generate. e.g. overflow
+	pub fn get_all_pool_token(pool_index: InvestingPoolIndex, epoch: u128) -> Option<Vec<AssetId>> {
```

Comment:
we create a series of tokens, so if user claimed one epoch reward, it will be converted into one equal amount of another type, until epoch overlapping the ending epoch => leads to nothing.

We actually do not iterate through all asset ids in logic.

Reply:
I think we do, in `create_investing_pool`:

```
// Create all asset token categories
let asset_id_vec: Vec<AssetIdOf<T>> =
InvestingPoolAssetIdGenerator::get_all_pool_token(pool_id, setting.epoch)
	.ok_or(ArithmeticError::Overflow)?;
for i in asset_id_vec.iter() {
<T::Fungibles as FsCreate<<T as frame_system::Config>::AccountId>>::create(
	*i,
	admin.clone(),
	true,
	One::one(),
)?;
}
```

> we create a series of tokens, so if user claimed one epoch reward, it will be converted into one equal amount of another type, until epoch overlapping the ending epoch => leads to nothing.

Maybe I still don't get the requirements, probably we need a sync